### PR TITLE
readd peer

### PR DESCRIPTION
**Pull request by tiancaiamao**
@siddontang @qiuyesuifeng 
this PR just add peer back.
I'll continue the work and fix https://github.com/pingcap/tikv/issues/528 in next PR.

### CODE REVIEW

File: src/bin/tikv-server.rs [hunk 1]
```diff
@@ -71,8 +71,8 @@ fn get_string_value<F>(short: &str,
 fn initial_log(matches: &Matches, config: &toml::Value) {
     let level = get_string_value("L",
                                  "server.log-level",
-                                 &matches,
-                                 &config,
+                                 matches,
```
Comment (siddontang) [inline src/bin/tikv-server.rs]: these change should be in another PR.

Comment (tiancaiamao) [inline src/bin/tikv-server.rs]: as we updated rust and clippy, compile output has many warning...
this is a small change and don't the affect logic

File: Cargo.lock [hunk 2]
```diff
@@ -4,17 +4,17 @@ version = "0.0.1"
 dependencies = [
  "byteorder 0.5.1 (registry+https://github.com/rust-lang/crates.io-index)",
  "bytes 0.3.0 (git+https://github.com/carllerche/bytes)",
- "clippy 0.0.65 (registry+https://github.com/rust-lang/crates.io-index)",
+ "clippy 0.0.66 (registry+https://github.com/rust-lang/crates.io-index)",
```
Comment (siddontang) [inline Cargo.lock]: why update all dependencies here? Only kvproto is enough. 
You should use another PR to update others.

**Comment by siddontang** [comment]
PTAL @BusyJay

File: Cargo.lock (hunk 2 above)
Comment (tiancaiamao) [inline Cargo.lock]: I don't know...
Cargo.lock is updated by cargo automatically, I just `make`

Comment (siddontang) [inline Cargo.lock]: `cargo update -p kvproto`

Comment (siddontang) [inline Cargo.lock]: make won't update the dependencies.

File: src/bin/tikv-server.rs (hunk 1 above)
Comment (siddontang) [inline src/bin/tikv-server.rs]: one PR does one thing, even the change is little.

Comment (tiancaiamao) [inline src/bin/tikv-server.rs]: okay...but I don't like PR with compile warning
this time it turn out to be a mistake with wrong clippy version

Comment (siddontang) [inline src/bin/tikv-server.rs]: If you meet compile warning later, please fire a PR to fix first, not combining it with other PRs.

Comment (siddontang) [inline src/bin/tikv-server.rs]: Btw, most of time, we won't meet these errors. After we update the rust and clippy, we must guarantee passing compile first.

**Comment by siddontang** [comment]
LGTM

File: src/raftstore/errors.rs [hunk 3]
```diff
@@ -39,9 +39,9 @@ quick_error!{
             description("region has not been initialized yet.")
             display("region {} not initialized yet", region_id)
         }
-        NotLeader(region_id: u64, leader_store_id: Option<u64>) {
+        NotLeader(region_id: u64, leader:Option<metapb::Peer>) {
```
Comment (BusyJay) [inline src/raftstore/errors.rs]: There should be a space after `:`.

Comment (tiancaiamao) [inline src/raftstore/errors.rs]: I can't understand why my rustfmt doesn't guarantee the format here.... Orz

File: src/raftstore/store/store.rs [hunk 4]
```diff
@@ -350,23 +362,27 @@ impl<T: Transport, C: PdClient> Store<T, C> {
     fn on_ready_change_peer(&mut self,
                             region_id: u64,
                             change_type: ConfChangeType,
-                            store_id: u64) {
+                            peer: metapb::Peer) {
         // We only care remove itself now.
-        if change_type == ConfChangeType::RemoveNode && store_id == self.store_id() {
-            warn!("destroy peer {:?} for region {}", store_id, region_id);
+        if change_type == ConfChangeType::RemoveNode && peer.get_store_id() == self.store_id() {
```
Comment (BusyJay) [inline src/raftstore/store/store.rs]: We  should check `peer_id` instead.

File: src/raftstore/store/util.rs [hunk 5]
```diff
@@ -20,20 +20,31 @@ use kvproto::raftpb::{self, ConfChangeType};
...
-    if let Some(index) = region.get_store_ids()
-                               .iter()
-                               .position(|&id| id == store_id) {
+    None
+}
 
-        region.mut_store_ids().remove(index);
-        return true;
+pub fn remove_peer(region: &mut metapb::Region, store_id: u64) -> Option<metapb::Peer> {
+    match region.get_peers()
```
Comment (BusyJay) [inline src/raftstore/store/util.rs]: use `map` instead.

File: src/raftstore/store/store.rs (hunk 4 above)
Comment (BusyJay) [inline src/raftstore/store/store.rs]: Seems we should check `peer_id` at L312 instead.

**Comment by ngaut** [comment]
PTAL @BusyJay @siddontang

**Comment by BusyJay** [comment]
LGTM

**Comment by ngaut** [comment]
LGTM

**Comment by siddontang** [comment]
I still persist on that you should remove unnecessary codes from this PR, re-fire in another and add more tests for this added feature.

**Comment by tiancaiamao** [comment]
remove the stale peer when we encounter a newer peer id with same store id, this can be viewed as part of **Peer** related change.
while I've already push a branch which **Focus on** how to remove dead peer, it don't has much to do with **Peer**, just querying Pd and get region info to validate it's legality~

**Comment by siddontang** [comment]
Do one thing in one PR. 

This PR is just for adding peer again, not fixing any other issue, and where is the corresponding test cases for this fix? 

/cc @qiuyesuifeng

**Comment by qiuyesuifeng** [comment]
I agree with @siddontang.
It is better to only add peer in this PR and fix compile warnings in another one.
Though this may be not a big deal, but it is a good habit for us.

**Comment by siddontang** [comment]
LGTM
